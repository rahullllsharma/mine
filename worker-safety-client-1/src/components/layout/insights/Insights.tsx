import type { ReactNode } from "react";
import cx from "classnames";
import EmptyContent from "@/components/emptyContent/EmptyContent";
import TabsLight from "@/components/shared/tabs/light/TabsLight";
import { useTenantStore } from "@/store/tenant/useTenantStore.store";
import PageLayout from "../pageLayout/PageLayout";

type InsightsProps = {
  filters?: ReactNode;
  charts?: ReactNode[];
  emptyChartsTitle: string;
  emptyChartsDescription?: string;
  onTabChange: (index: number) => void;
  hasData?: boolean;
};

const EmptyChartContent = ({
  title,
  description = "Try adjusting your filters, to return more data",
}: {
  title: string;
  description?: string;
}): JSX.Element => (
  <div className="bg-white rounded-xl flex flex-1 items-center justify-center min-h-[130px] max-h-48">
    <EmptyContent title={title} description={description} />
  </div>
);

export default function Insights({
  filters,
  charts = [],
  emptyChartsTitle,
  emptyChartsDescription,
  onTabChange,
  hasData = true,
}: InsightsProps): JSX.Element {
  const { workPackage } = useTenantStore(state => state.getAllEntities());
  const tabOptions = ["Portfolio", workPackage.label];

  return (
    <PageLayout className="responsive-padding-x bg-white" showExpand={true}>
      <div className="h-full hidden lg:block">
        <TabsLight options={tabOptions} onSelect={onTabChange} />

        <div className="flex flex-wrap  bg-white rounded-xl p-6 mt-5 ">
          <div className="lg:w-full pr-4 pl-4">
            <div className="sticky top-0">{filters}</div>
          </div>
        </div>

        <div className="flex mt-6 pb-6 gap-x-8 ">
          {charts.length > 0 && (
            <div
              className={cx("flex flex-col gap-y-8 flex-1 my-14", {
                // we need to 'render' the chart comps even when they don't have
                // data, so they can make more requests when the filters change
                ["relative"]: hasData,
                ["hidden"]: !hasData,
              })}
            >
              {charts.map((chart, index) => (
                <section
                  key={index}
                  className="bg-white p-6 rounded-xl overflow-hidden"
                >
                  {chart}
                </section>
              ))}
            </div>
          )}
          {(charts.length === 0 || !hasData) && (
            <EmptyChartContent
              title={emptyChartsTitle}
              description={emptyChartsDescription}
            />
          )}
        </div>
      </div>
      <div className="h-2/3 w-full flex items-center lg:hidden">
        <EmptyContent
          title="Your screen size is not large enough to support this feature"
          description="Please increase your screen size to access the content"
        />
      </div>
    </PageLayout>
  );
}
