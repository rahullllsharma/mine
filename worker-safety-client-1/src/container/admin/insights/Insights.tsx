import type { Insight, InsightFormInputs } from "@/types/insights/Insight";
import type { OnDragEndResponder } from "react-beautiful-dnd";
import { useQueryClient } from "@tanstack/react-query";
import { useCallback, useContext, useEffect, useState } from "react";
import { differenceWith, fromPairs, isEqual, toPairs } from "lodash-es";
import { BodyText, SectionHeading } from "@urbint/silica";
import { QUERY_KEYS } from "@/api/restApi";
import EmptyContent from "@/components/emptyContent/EmptyContent";
import ButtonPrimary from "@/components/shared/button/primary/ButtonPrimary";
import ToastContext from "@/components/shared/toast/context/ToastContext";
import useRestMutation from "@/hooks/useRestMutation";
import useRestQuery from "@/hooks/useRestQuery";
import AddInsightModal from "./AddInsightModal";
import DeleteInsightConfirmationModal from "./DeleteInsightConfirmationModal";
import InsightsTable from "./InsightsTable";

const Insights = () => {
  const [isAddInsightModalOpen, setIsAddInsightModalOpen] = useState(false);
  const [selectedId, setSelectedId] = useState<string | null>(null);
  const [toDeleteId, setToDeleteId] = useState<string | null>(null);
  const [isDeleteConfirmationModalOpen, setIsDeleteConfirmationModalOpen] =
    useState(false);

  const toastCtx = useContext(ToastContext);
  const queryClient = useQueryClient();

  const openAddInsightModal = () => {
    setIsAddInsightModalOpen(true);
  };

  const closeAddInsightModal = () => {
    setIsAddInsightModalOpen(false);
    setSelectedId(null);
  };

  const openDeleteConfirmationModal = () => {
    setIsDeleteConfirmationModalOpen(true);
  };

  const closeDeleteConfirmationModal = () => {
    setIsDeleteConfirmationModalOpen(false);
    setToDeleteId(null);
  };

  const {
    data: insights,
    isLoading,
    error,
  } = useRestQuery<
    { data: { attributes: Insight; id: string }[] },
    { data: Insight[] }
  >({
    key: [QUERY_KEYS.INSIGHTS],
    endpoint: "/insights",
    axiosConfig: {
      params: {
        "page[limit]": 100,
      },
    },
    queryOptions: {
      select(data) {
        return {
          data: data.data?.map(dt => ({ ...dt.attributes, id: dt.id })) || [],
        };
      },
    },
  });

  const {
    mutate: addInsight,
    isLoading: isAdding,
    error: addError,
  } = useRestMutation<InsightFormInputs>({
    endpoint: "/insights",
    method: "post",
    dtoFn: data => ({
      data: {
        attributes: { ...data },
      },
      type: "insight",
    }),
    mutationOptions: {
      onSuccess: () => {
        toastCtx?.pushToast("success", "Insight added successfully");
        queryClient.refetchQueries({ queryKey: [QUERY_KEYS.INSIGHTS] });
      },
    },
  });

  const {
    mutate: updateInsight,
    isLoading: isUpdating,
    error: updateError,
  } = useRestMutation<{ data: Partial<InsightFormInputs>; id: string }>({
    endpoint: data => `/insights/${data.id}`,
    method: "put",
    dtoFn: data => ({
      data: {
        attributes: { ...data.data },
      },
      type: "insight",
    }),
    mutationOptions: {
      onSuccess: () => {
        toastCtx?.pushToast("success", "Insight updated successfully");
        queryClient.refetchQueries({ queryKey: [QUERY_KEYS.INSIGHTS] });
      },
    },
  });

  const {
    mutate: deleteInsight,
    isLoading: isDeleting,
    error: deleteError,
  } = useRestMutation<{ id: string }>({
    endpoint: data => `/insights/${data.id}`,
    method: "delete",
    mutationOptions: {
      onSuccess: () => {
        closeDeleteConfirmationModal();
        toastCtx?.pushToast("success", "Insight deleted successfully");
        queryClient.refetchQueries({ queryKey: [QUERY_KEYS.INSIGHTS] });
      },
    },
  });

  const { mutate: reorderInsights, error: reorderError } = useRestMutation<{
    ids: string[];
  }>({
    endpoint: "/insights/reorder/",
    method: "put",
    dtoFn: data => data.ids,
    mutationOptions: {
      // Optimistic update
      onMutate: async ({ ids: newInsightIds }) => {
        // Cancel any outgoing refetches
        // (so they don't overwrite our optimistic update)
        await queryClient.cancelQueries({ queryKey: [QUERY_KEYS.INSIGHTS] });

        // Snapshot the previous value
        const previousInsightsData = queryClient.getQueryData([
          QUERY_KEYS.INSIGHTS,
        ]) as {
          data: { attributes: Insight; id: string }[];
        };

        // Optimistically update to the new value
        queryClient.setQueryData([QUERY_KEYS.INSIGHTS], () => ({
          data: newInsightIds.map(id =>
            previousInsightsData.data.find(insight => insight.id === id)
          ),
        }));

        // Return a context object with the snapshotted value
        return { previousInsightsData };
      },
      // If the mutation fails,
      // use the context returned from onMutate to roll back
      onError: (err, _, context) => {
        queryClient.setQueryData(
          [QUERY_KEYS.INSIGHTS],
          (
            context as {
              previousInsightsData: {
                data: { attributes: Insight; id: string }[];
              };
            }
          ).previousInsightsData
        );
      },
      onSettled: () => {
        queryClient.refetchQueries({ queryKey: [QUERY_KEYS.INSIGHTS] });
      },
    },
  });

  const handleAddInsightSubmit = (data: InsightFormInputs) => {
    if (selectedId) {
      const changes = fromPairs(
        differenceWith(
          toPairs(data),
          toPairs(
            insights?.data.find(insight => insight.id === selectedId) || {}
          ),
          isEqual
        )
      );
      updateInsight({ data: { ...changes }, id: selectedId });
      return;
    }

    addInsight(data);
  };

  const onEdit = useCallback((id: string) => {
    setSelectedId(id);
    setTimeout(() => {
      openAddInsightModal();
    }, 100);
  }, []);

  const onDeleteRequest = useCallback((id: string) => {
    setToDeleteId(id);
    closeAddInsightModal();
    openDeleteConfirmationModal();
  }, []);

  const onDelete = useCallback(
    (id: string) => {
      deleteInsight({ id });
    },
    [deleteInsight]
  );

  const onReOrder: OnDragEndResponder = useCallback(
    result => {
      const { source, destination, draggableId } = result;

      if (!destination) {
        return;
      }

      if (source.index === destination.index) {
        return;
      }

      if (!insights) {
        return;
      }

      const newInsights = insights.data.map(insight => insight.id) || [];
      newInsights.splice(source.index, 1);
      newInsights.splice(destination.index, 0, insights.data[+draggableId]?.id);
      reorderInsights({ ids: newInsights });
    },
    [insights, reorderInsights]
  );

  useEffect(() => {
    const errorMessage =
      error || addError || updateError || deleteError || reorderError;
    if (errorMessage) {
      toastCtx?.pushToast("error", errorMessage);
    }

    // eslint-disable-next-line react-hooks/exhaustive-deps
  }, [error, addError, updateError, deleteError, reorderError]);

  return (
    <div className="flex flex-col flex-1 overflow-hidden">
      <header className="text-neutral-shade-100 flex justify-between items-start mb-8">
        <div>
          <SectionHeading className="text-xl font-semibold">
            Insights
          </SectionHeading>
          <BodyText className="text-base">
            Create and modify Insights for your organization
          </BodyText>
        </div>
        <ButtonPrimary
          label="Add Insights"
          size="lg"
          onClick={openAddInsightModal}
        />
      </header>
      <div className="flex flex-1 overflow-hidden">
        {error ? (
          <EmptyContent
            title="Error getting data"
            description="Please contact the engineering team to help troubleshoot the issues"
          />
        ) : (
          <InsightsTable
            data={insights?.data || []}
            onEdit={onEdit}
            onDelete={onDeleteRequest}
            onReOrder={onReOrder}
            isLoading={isLoading}
          />
        )}
      </div>
      <AddInsightModal
        isOpen={isAddInsightModalOpen}
        onClose={closeAddInsightModal}
        onSubmit={handleAddInsightSubmit}
        defaultValues={
          selectedId
            ? insights?.data.find(insight => insight.id === selectedId)
            : undefined
        }
        onDelete={onDeleteRequest}
        isSubmitting={isAdding || isUpdating}
      />
      <DeleteInsightConfirmationModal
        isOpen={isDeleteConfirmationModalOpen}
        onClose={closeDeleteConfirmationModal}
        onSubmit={onDelete}
        id={toDeleteId}
        isDeleting={isDeleting}
      />
    </div>
  );
};

export default Insights;
