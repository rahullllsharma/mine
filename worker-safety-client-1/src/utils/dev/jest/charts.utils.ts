import type { RiskLevelByDate } from "@/components/charts/dailyRiskHeatmap/types";
import { act } from "@testing-library/react";
import { RiskLevel } from "@/components/riskBadge/RiskLevel";

export function chartRenderHelper(): () => void {
  // set up a listener so we can force the width/height
  // this is required to get the chart to render in its ResponsiveWrapper
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  let listener: ((rect: any) => void) | undefined;
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  (window as any).ResizeObserver = class ResizeObserver {
    // eslint-disable-next-line @typescript-eslint/no-explicit-any
    constructor(ls: any) {
      listener = ls;
    }
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    observe() {}
    // eslint-disable-next-line @typescript-eslint/no-empty-function
    disconnect() {}
  };

  function triggerResize(): void {
    act(() => {
      listener &&
        listener([
          {
            contentRect: {
              x: 0,
              y: 0,
              width: 800,
              height: 800,
              top: 100,
              bottom: 0,
              left: 100,
              right: 0,
            },
          },
        ]);
    });
  }
  return triggerResize;
}

export const sampleHeatmapDates = {
  startDate: "2022-01-29",
  endDate: "2022-02-03",
};

// consumed by the DailyRiskHeatmap
export const sampleRiskLevelByDate: RiskLevelByDate[] = [
  {
    date: "2022-01-29",
    riskLevel: RiskLevel.HIGH,
  },
  {
    date: "2022-01-30",
    riskLevel: RiskLevel.MEDIUM,
  },
  {
    date: "2022-01-31",
    riskLevel: RiskLevel.UNKNOWN,
  },
  {
    date: "2022-02-01",
    riskLevel: RiskLevel.LOW,
  },
  {
    date: "2022-02-02",
    riskLevel: RiskLevel.RECALCULATING,
  },
  {
    date: "2022-02-03",
    riskLevel: RiskLevel.UNKNOWN,
  },
];
