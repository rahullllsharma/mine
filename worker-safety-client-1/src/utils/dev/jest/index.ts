export * from "./intl.utils";
export * from "./testing.utils";
export * from "./charts.utils";
