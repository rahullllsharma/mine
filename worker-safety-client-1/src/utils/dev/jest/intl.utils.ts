/** Mock the default locale when using the Intl.DateTimeFormat class */
export function mockIntlDateTimeFormatLocale(locale: string): void {
  const { DateTimeFormat } = Intl;
  jest
    .spyOn(global.Intl, "DateTimeFormat")
    .mockImplementation((_, options) => new DateTimeFormat(locale, options));
}
