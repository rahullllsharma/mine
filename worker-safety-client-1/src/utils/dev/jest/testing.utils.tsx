/* istanbul ignore file */
import type { ReactNode } from "react";
import type { PropsWithClassName } from "@/types/Generic";
import { fireEvent, render, screen } from "@testing-library/react";
import { FormProvider, useForm } from "react-hook-form";
import { useTenantStore } from "@/store/tenant/useTenantStore.store";
import { TenantMock } from "@/store/tenant/utils/tenantMock";

export const formTemplate = (
  component: ReactNode,
  defaultValues: { [key: string]: any } = {}
) => {
  function Wrapper({ children }: PropsWithClassName): JSX.Element {
    const methods = useForm({ defaultValues });
    return <FormProvider {...methods}>{children}</FormProvider>;
  }
  return <Wrapper>{component}</Wrapper>;
};

export const formRender = (
  component: ReactNode,
  defaultValues: { [key: string]: any } = {}
) => {
  return render(formTemplate(component, defaultValues));
};

// to overcome react-select limitation for testing, of not responding to click events
export const openSelectMenu = (role = "combobox", name?: RegExp): void => {
  const inputElement = screen.getByRole(role, { name });
  fireEvent.keyDown(inputElement, { key: "ArrowDown", code: "ArrowDown" });
};

export const mockTenantStore = (mock = TenantMock) => {
  const { setTenant } = useTenantStore.getState();
  setTenant(mock);
};
