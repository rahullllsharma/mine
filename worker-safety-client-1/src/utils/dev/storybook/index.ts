export * from "./WrapperForm.utils";
export * from "./WrapperModal.utils";
export * from "./WrapperScroll.utils";
