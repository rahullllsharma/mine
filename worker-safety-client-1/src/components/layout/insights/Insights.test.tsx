import { render, screen } from "@testing-library/react";
import { mockTenantStore } from "@/utils/dev/jest";
import Insights from "./Insights";

const mockTabChange = jest.fn();

describe(Insights.name, () => {
  mockTenantStore();
  describe("when it renders", () => {
    it("should match the snapshot", () => {
      const { asFragment } = render(
        <Insights emptyChartsTitle="" onTabChange={mockTabChange} />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe("when contains a filter element", () => {
    it("should render the filter element", () => {
      render(
        <Insights
          filters={<div>Lorem ipsum</div>}
          emptyChartsTitle=""
          onTabChange={mockTabChange}
        />
      );
      screen.getByText(/lorem ipsum/i);
    });
  });

  describe("when contains a chart element", () => {
    it("should render the chart element", () => {
      render(
        <Insights
          charts={[<div key={1}>Lorem ipsum</div>]}
          emptyChartsTitle=""
          onTabChange={mockTabChange}
        />
      );
      screen.getByText(/lorem ipsum/i);
    });
  });

  describe("when no charts are passed", () => {
    it("should render empty content message", () => {
      render(
        <Insights
          filters={<div>Lorem ipsum</div>}
          emptyChartsTitle="There are no charts available"
          onTabChange={mockTabChange}
        />
      );
      screen.getByText(/there are no charts available/i);
    });
  });
});
