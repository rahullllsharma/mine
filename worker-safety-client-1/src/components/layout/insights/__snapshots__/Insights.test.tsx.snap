// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Insights when it renders should match the snapshot 1`] = `
<DocumentFragment>
  <section
    class="relative mb-8"
  >
    <button
      class="text-xl text-neutral-shade-75 disabled:text-neutral-shade-38 disabled:cursor-not-allowed absolute top-0 right-4"
    >
      <i
        aria-hidden="true"
        class="ci-expand"
      />
    </button>
  </section>
  <main
    class="flex flex-col bg-brand-gray-10 overflow-y-auto lg:overflow-hidden w-full"
  >
    <section
      class="flex flex-col overflow-y-auto h-full responsive-padding-y responsive-padding-x bg-white"
      data-testid="page-layout"
      id="page-layout"
    >
      <div
        class="h-full hidden lg:block"
      >
        <div
          aria-orientation="horizontal"
          class="flex"
          role="tablist"
        >
          <button
            aria-selected="true"
            class="min-w-0 z-10 outline-none"
            id="headlessui-tabs-tab-1"
            role="tab"
            tabindex="0"
            type="button"
          >
            <div
              class="flex items-center px-2.5 py-3 text-neutral-shade-75 font-semibold active:bg-neutral-shade-3 border-b-4 border-brand-urbint-40 border-solid z-10"
            >
              <span
                class="text-base mx-1 truncate"
              >
                Portfolio
              </span>
            </div>
          </button>
          <button
            aria-selected="false"
            class="min-w-0 z-10 outline-none"
            id="headlessui-tabs-tab-2"
            role="tab"
            tabindex="-1"
            type="button"
          >
            <div
              class="flex items-center px-2.5 py-3 text-neutral-shade-75 font-semibold active:bg-neutral-shade-3 border-b-4 opacity-75 border-transparent"
            >
              <span
                class="text-base mx-1 truncate"
              >
                Project
              </span>
            </div>
          </button>
        </div>
        <hr
          class="relative bg-neutral-shade-18 z-1 border"
          style="bottom: 3px;"
        />
        <div
          class="flex flex-wrap bg-white rounded-xl p-6 mt-5 "
        >
          <div
            class="lg:w-full pr-4 pl-4"
          >
            <div
              class="sticky top-0"
            />
          </div>
        </div>
        <div
          class="flex mt-6 pb-6 gap-x-8 "
        >
          <div
            class="bg-white rounded-xl flex flex-1 items-center justify-center min-h-[130px] max-h-48"
          >
            <div
              class="flex flex-col items-center w-full text-center px-6"
            >
              <p
                class="text-neutral-shade-100 text-xl font-semibold mb-2"
              />
              <p
                class="text-brand-gray-70 text-sm max-w-md"
              >
                Try adjusting your filters, to return more data
              </p>
            </div>
          </div>
        </div>
      </div>
      <div
        class="h-2/3 w-full flex items-center lg:hidden"
      >
        <div
          class="flex flex-col items-center w-full text-center px-6"
        >
          <p
            class="text-neutral-shade-100 text-xl font-semibold mb-2"
          >
            Your screen size is not large enough to support this feature
          </p>
          <p
            class="text-brand-gray-70 text-sm max-w-md"
          >
            Please increase your screen size to access the content
          </p>
        </div>
      </div>
    </section>
  </main>
</DocumentFragment>
`;
